### DISCUSSION TITLE
Making PathPlannerTrajectory::Waypoint public?

### DISCUSSION THREAD

**Discussion by FarkasJoseph** [mjansen4857/pathplanner#169] [General]
My team ended up having a use case for trajectories where the normal generatePath functions in the PathPlanner namespace were not enough. Instead, we ended up needing to make our own waypoints (not pathpoints) and used the constructor for PathPlannerTrajectory (takes in waypoints, not pathpoints). We are wondering if it would be possible to have the waypoint class be public (we just made the change in our cache). I'm not sure what your preferred solution is for this (the easiest would be just making it public, but it might also be better to move it into its own namespace/file), so let me know what you think. I can PR this if you approve.

**Comment by mjansen4857** [comment]
What is the use case that path point doesn't work for? The waypoint class isn't public since it needs the control point positions set in field coordinates to determine heading and the shape of the path, which is super unintuitive for the user. PathPoint is supposed to function as an abstraction that lets you just specify a position and heading.

I think the better solution would be to support this use case with path point/generate path methods unless its something really specific.

**Reply by FarkasJoseph** [reply]
Basically we're trying to make trajectories with finer control over how to enter/exit a point. The best way we found to accomplish this was to modify the magnitude of the heading vector, aka the length of the spline handle, just like you can in the GUI.

**Reply by FarkasJoseph** [reply]
Do you have any thoughts?

**Reply by mjansen4857** [reply]
Sorry, discussions get buried in my email since new activity for them isn't very visible on github. I think for this use case it makes sense to have a configurable magnitude factor in the pathpoint class. Currently, the distance of the control point from the anchor point is 1/3 the distance between the next anchor point. This could be made a configurable value between 0-1 and could then be used for finer control without having to manually set control point positions.

**Reply by FarkasJoseph** [reply]
Is there any reason it has to be limited between 0-1?

**Reply by mjansen4857** [reply]
No, I suppose not. It should at least be greater than 0 though. Only thing is you need to be careful when using this since overlapping control points can make really janky paths. Its easy to see this in the GUI, but through code, especially when generating on the fly its really hard to prevent. Making sure that value is below 0.5 would prevent a lot of this jank, but limits you significantly in places that would normally be fine.

**Reply by FarkasJoseph** [reply]
Are you referring to this kind of jank? 
![image](https://user-images.githubusercontent.com/16584585/206339811-711c4797-8092-4b4a-ba67-a055c3926aac.png)

**Reply by mjansen4857** [reply]
Exactly. 

I suppose another solution would be to just use a length in meters instead of some factor. That way you could have a consistent control point for, say, and end point at a loading station regardless of where the waypoint before that it.

**Reply by FarkasJoseph** [reply]
That sounds good
